### ISSUE TITLE
Opacity widget always assumes compositing

### ISSUE THREAD

**Issue by flar** [flutter/flutter#102179]
Most widgets that modify the rendering of their children, or rather the RenderObject associated with those widgets, have code that detects if the children need compositing. If the children need compositing then they inject themselves into the tree as a layer. If the children do not need compositing then they simply issue rendering commands (usually along with a save() or saveLayer()) to set the rendering modification into the current picture/display list.

See the way that the `pushTransform` method tests compositing and changes the way it applies itself to rendering the tree:
https://github.com/flutter/flutter/blob/b9f1a907b1aa5f13d0443cecb7e9f851b153c11f/packages/flutter/lib/src/rendering/object.dart#L562

OpacityLayer does not have the same tests and always adds a layer, it could simply emit a saveLayer with a Paint object that has the opacity in it if the children don't need compositing.

**Comment by ColdPaleLight** [comment]
Hi, @flar 
I'm not sure if this is a bug or just by design.

Let's first look at the implementation of `RenderTransform`. `RenderTransform` calls `pushTransform` in the `paint` method and passes `needsCompositing` as an argument.
https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/proxy_box.dart#L2404-L2410

And getter `needsCompositing` returns `_needsCompositing`, which is always true if `isRepaintBoundary` or `alwaysNeedsCompositing` returns `true`

https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/object.dart#L2211-L2214

https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/object.dart#L1238

https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/object.dart#L2226-L2227

For `RenderTransform`, its `isRepaintBoundary` returns `false`, and `alwaysNeedsCompositing` is implemented as follows
https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/proxy_box.dart#L2273

So if `RenderTransform` does not set `filterQuality`, it is possible that compositing is not required.


Let's take another look at `RenderOpacity`, it's `alwaysNeedsCompositing` is implemented as follows
https://github.com/flutter/flutter/blob/51bcdb9407211ccaf0616533fe1f9eb0e8f04fca/packages/flutter/lib/src/rendering/proxy_box.dart#L846

That is to say, as long as alpha is not 0, then it needs compositing, that is, in most cases, it needs compositing.

In this PR https://github.com/flutter/flutter/pull/101952, `RenderOpacity`'s `isRepaintBoundary` and `alwaysNeedsCompositing` have made some changes, but it does not affect the previous conclusion.
``` diff
-    bool get alwaysNeedsCompositing => child != null && (_alpha > 0);
+    bool get isRepaintBoundary => child != null && (_alpha > 0);
```

To sum up, `Opacity` widget always assumes compositing looks by design. please let me know if I missed something, thanks you.

cc @jonahwilliams @dnfield

**Comment by jonahwilliams** [comment]
This is true, however there is nothing in that proposal that requires compositing - it simply mirrors the existing behavior. The intention is that if we are already compositing, then we might as well be a repaint boundary too.

Having said that, I have some concerns on just how often an opacity wouldn't need compositing. Almost by design, many layer types like the annotated region or physical model force composition. Unwinding that is possible but will be difficult

**Comment by flar** [comment]
We need to unwind these things. Don't give up hope.

@ColdPaleLight the distinction is "can this layer do its thing without adding a layer".

When a child needs compositing then any state which a RO wants to apply to "all children" will not survive the break in the Picture that happens when that child interrupts the picture with a layer. So, the compositing flag ripples up the tree, splitting it.

Both Transform and Opacity fall into this category in that they are going to impose non-transferable state on their children. So, if a child needs compositing then both must add a layer.

Transform also needs to add a layer if the filterQuality is there because it will accomplish its goal using a filter.

Opacity has no other reason to add a layer. It can always just:

```
  Paint p = Paint()..color = Color(alpha << 24);
  context.canvas.saveLayer(..., p);
  {paint children}
  context.canvas.restore();
```

But, currently OpacityLayer never takes that option.

**Comment by jonahwilliams** [comment]
Makes sense! I suspect that I'm not the only one interested in taking apart the physical model layer anyway. I do think the annotated region will be tricky, but that one is my fault in the first place

**Comment by dnfield** [comment]
We should just remove physical model: https://github.com/flutter/flutter/issues/91852

Those only ever existed as layers for Fuchsia, and Fuchsia doesn't care about that anymore.

**Comment by jonahwilliams** [comment]
It wasn't actually that hard to update pml to stop requiring compositing. What google3 goldens will show is another story...

**Comment by jonahwilliams** [comment]
A change to remove compositing from opacity was reverted as it regressed the opacity peephole benchmarks significantly: https://github.com/flutter/flutter/pull/105489
